r/tests/testthat/test-dplyr-slice.R

```diff
@@ -0,0 +1,192 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+library(dplyr, warn.conflicts = FALSE)
+
+tbl <- example_data
+
+test_that("slice_head/tail, ungrouped", {
+  # head/tail are not deterministic in Arrow because data is unordered
+  # so we can't assert identical to dplyr, just assert right number of rows
+  tab <- arrow_table(tbl)
+  expect_equal(
+    tab %>%
+      slice_head(n = 5) %>%
+      nrow(),
+    5
+  )
+  expect_equal(
+    tab %>%
+      slice_tail(n = 5) %>%
+      nrow(),
+    5
+  )
+
+  expect_equal(
+    tab %>%
+      slice_head(prop = .25) %>%
+      nrow(),
+    2
+  )
+  expect_equal(
+    tab %>%
+      slice_tail(prop = .25) %>%
+      nrow(),
+    2
+  )
+})
+
+test_that("slice_min/max, ungrouped", {
+  # with_ties must be FALSE
+  tab <- arrow_table(tbl)
+  expect_error(
+    tab %>% slice_max(int, n = 5),
+    "with_ties = TRUE"
+  )
+  expect_error(
+    tab %>% slice_min(int, n = 5),
+    "with_ties = TRUE"
+  )
+  compare_dplyr_binding(
+    .input %>%
+      slice_max(int, n = 4, with_ties = FALSE) %>%
+      collect(),
+    tbl
+  )
+  compare_dplyr_binding(
+    .input %>%
+      slice_min(int, n = 4, with_ties = FALSE) %>%
+      collect(),
+    tbl
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      slice_max(int, prop = .25, with_ties = FALSE) %>%
+      collect(),
+    tbl
+  )
+  compare_dplyr_binding(
+    .input %>%
+      slice_min(int, prop = .25, with_ties = FALSE) %>%
+      collect(),
+    tbl
+  )
+})
+
+test_that("slice_sample, ungrouped", {
+  tab <- arrow_table(tbl)
+  expect_error(
+    tab %>% slice_sample(replace = TRUE),
+    "Sampling with replacement"
+  )
+  expect_error(
+    tab %>% slice_sample(weight_by = dbl),
+    "weight_by"
+  )
+
+  # Because this is random (and we only have 10 rows), try several times
+  for (i in 1:10) {
+    sampled_prop <- tab %>%
+      slice_sample(prop = .2) %>%
+      collect() %>%
+      nrow()
+    if (sampled_prop == 2) break
+  }
+  expect_equal(sampled_prop, 2)
```

Comment:
It's implemented as `filter(runif(nrow) < prop)`, and since that's random, you won't always get an exact number or rows.

```
> tab <- mtcars %>%
+   arrow_table() %>%
+   slice_sample(prop = .2)
> 
> for (i in 1:10) print(nrow(collect(tab)))
[1] 3
[1] 12
[1] 6
[1] 4
[1] 4
[1] 4
[1] 4
[1] 5
[1] 4
[1] 7
> .2 * nrow(mtcars)
[1] 6.4
```